src/main/appender.cpp

```diff
@@ -375,6 +411,65 @@ void Appender::FlushInternal(ColumnDataCollection &collection) {
 	context->Append(*description, collection);
 }
 
+void Appender::AppendDefaultsToVector(Vector &result, idx_t vector_size, idx_t column, SelectionVector &sel,
+                                      idx_t count) {
+	sel.Verify(count, vector_size);
+	if (count > STANDARD_VECTOR_SIZE) {
+		throw InvalidInputException(
+		    "Please provide a SelectionVector that is at most STANDARD_VECTOR_SIZE (%d) in size", STANDARD_VECTOR_SIZE);
+	}
+	if (vector_size > STANDARD_VECTOR_SIZE) {
+		throw InvalidInputException("Please provide a Vector that is at most STANDARD_VECTOR_SIZE (%d) in size",
+		                            STANDARD_VECTOR_SIZE);
+	}
+	if (column >= types.size()) {
+		throw InvalidInputException(
+		    "Could not append DEFAULT value of column %d, out of range index, table only has %d columns", column,
+		    types.size());
+	}
+	auto &type = types[column];
+	if (result.GetType() != type) {
+		throw InvalidInputException("Type mismatch, column has type: %s, while the provided Vector has type %s",
+		                            type.ToString(), result.GetType().ToString());
+	}
+	auto &executor = *expression_executor;
+
+	// Create an empty chunk to tell the ExpressionExecutor how many tuples to execute for
+	DataChunk empty_chunk;
+	empty_chunk.SetCardinality(count);
+	executor.SetChunk(empty_chunk);
+
+	Vector intermediate(type, count);
+	context->RunFunctionInTransaction([&]() { executor.ExecuteExpression(column, intermediate); });
+	for (idx_t i = 0; i < count; i++) {
+		auto temp_idx = i;
+		auto result_idx = sel.get_index(i);
+		auto value = intermediate.GetValue(temp_idx);
+		result.SetValue(result_idx, value);
+	}
+}
+
+void Appender::AppendDefault() {
+	auto &default_expr = bound_defaults[column];
+	if (default_expr->type == ExpressionType::VALUE_CONSTANT) {
+		// Fast path: NULL or constant-folded
+		auto &bound_constant = default_expr->Cast<BoundConstantExpression>();
+		Append(bound_constant.value);
+		return;
+	}
+
+	auto &type = types[column];
+	Vector result(type, 1);
+	auto &executor = *expression_executor;
+	// The executor is initialized with expressions for every column, even though only some are used
+	// this makes it so that we can just use the 'column' index
+
+	// FIXME: we could optimize this further by delaying this until Flush is called, then doing it for the entire chunk
+	// once
+	context->RunFunctionInTransaction([&]() { executor.ExecuteExpression(column, result); });
```

Comment:
We need to keep the transaction open to support this - otherwise e.g. references to catalog entries (like sequences) can become invalid. Maybe we can just not support it entirely for now and only allow foldable expressions to be executed here?

I think the "nice" way of supporting this is to only execute the default values (that are non-foldable) in the flush, as a transaction is already opened in `ClientContext::Append`. Maybe we can keep track of e.g. a selection vector per column of which default values should be computed? But maybe this can also be done in a follow-up PR to keep this one limited in scope.